oslo_vmware/rw_handles.py

```diff
@@ -706,3 +712,48 @@ def close(self):
 
     def __str__(self):
         return "Image read handle"
+
+
+class UrlPullHandle(object):
+    """Base class for a URL from where VMware can download VMDKs
+
+    This kind of handles are being used in conjunction with
+    HttpNfcLeasePullFromUrls_Task, converting an existing lease
+    to "Pull" mode.
+    """
+
+    def __init__(self):
+        self._url = None
```

Comment:
Personally, I prefer a programming style which doesn't allow for invalid states in an object.
So, no `is_valid` method and either:
- the variable is passed via the constructor, and validated here,
- or that the class is an abstract base class with `url` not implemented.

Either way, I think it would be good for  `ssl_thumbprint` uses `self.url()` instead of `self._url`, so that when an implementation overrides it, it also has an effect on `ssl_thumbprint`

Not sure though, which of those two options I prefer, they have both advantages and disadvantages.

Reply:
We need a way to know if the `UrlPullHandle` can be used and the data passed in is valid (i.e: we passed a direct_url that is not swift, thus we must fallback to proxying the image with `VmdkHandle`)

By removing `is_valid`, then I propose than `url()` can return `None`, case when the caller must fallback to other handle.

Reply:
> We need a way to know if the UrlPullHandle can be used and the data passed in is valid

Yes, and with my first proposal the way to know it, is by being able to construct such a handle. It is by definition valid.
Ergo:
```python
try:
  url_handle = rw_handles.SwiftUrlPullHandle(direct_url, context.get_auth_plugin().auth_token)
except ValueError:
  url_handle = None

if url_handle:
  imported_vm = image_pull_from_url(url_handle, **kwargs)
  LOG.debug("Downloaded image: %s from image direct URL as a stream "
                       "optimized file.")
  return imported_vm

...
```

Reply:
Fixed (I hope).